src/loki/reach.cc

```diff
@@ -110,17 +131,117 @@ directed_reach SimpleReach(const DirectedEdge* edge,
+
+  // if we didnt make the limit we should do an exact check for whichever directions we didnt meet
+  auto retry_direction = (direction & kOutbound && reach.outbound < max_reach ? kOutbound : 0) |
+                         (direction & kInbound && reach.inbound < max_reach ? kInbound : 0);
+  if (retry_direction) {
+    auto retry_reach = exact(edge, edge_id, max_reach, reader, costing, retry_direction);
+    reach.outbound = std::max(reach.outbound, retry_reach.outbound);
+    reach.inbound = std::max(reach.inbound, retry_reach.inbound);
+  }
 
   return reach;
 }
 
+directed_reach Reach::exact(const valhalla::baldr::DirectedEdge* edge,
+                            const GraphId edge_id,
+                            uint32_t max_reach,
+                            valhalla::baldr::GraphReader& reader,
+                            const std::shared_ptr<sif::DynamicCost>& costing,
+                            uint8_t direction) {
+  // can we even try to expand?
+  directed_reach reach{};
+  if (costing->GetEdgeFilter()(edge) == 0) {
+    return reach;
+  }
+
+  // fake up the input location
+  const baldr::GraphTile* tile = nullptr;
+  const auto* node = reader.GetEndNode(edge, tile);
+  auto ll = node->latlng(tile->header()->base_ll());
+  locations_.Mutable(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_ll()->set_lat(ll.second);
+  locations_.Mutable(0)->mutable_path_edges(0)->set_graph_id(edge_id);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lng(ll.first);
+  locations_.Mutable(0)->mutable_path_edges(0)->mutable_ll()->set_lat(ll.second);
+
+  // fake up the costing array
+  std::shared_ptr<sif::DynamicCost> costings[static_cast<int>(sif::TravelMode::kMaxTravelMode)];
+  costings[static_cast<int>(costing->travel_mode())] = costing;
+
+  // expand in the forward direction
+  if (direction | kOutbound) {
+    Clear();
+    Compute(locations_, reader, costings, costing->travel_mode());
+    reach.outbound = std::min(static_cast<uint32_t>(done_.size() - transitions_), max_reach);
+  }
+
+  // expand in the reverse direction
+  if (direction | kInbound) {
+    Clear();
+    ComputeReverse(locations_, reader, costings, costing->travel_mode());
+    reach.inbound = std::min(static_cast<uint32_t>(done_.size() - transitions_), max_reach);
+  }
+
+  return reach;
+}
+
+// callback fired when a node is expanded from, the node will be the end node of the previous label
+void Reach::ExpandingNode(baldr::GraphReader& graphreader,
+                          const baldr::GraphTile* tile,
+                          const baldr::NodeInfo* node,
+                          const sif::EdgeLabel& current,
+                          const sif::EdgeLabel* previous) {
+  // compute the nodes id
+  GraphId node_id = tile->header()->graphid();
+  node_id.set_id(node - tile->node(0));
+  // mark the node and if its a new one then mark its doppelgänger on the other levels
+  if (done_.insert(node_id).second) {
```

Comment:
when the exact expansion expands from a node we mark it in the done set and if it wasnt already seen we also mark any transitions that occur there so they dont get double counted.